packages/core-utils/src/watcher.ts

```diff
@@ -73,22 +73,31 @@ export class Watcher {
     msgHash: string,
     pollForPending: boolean = true
   ): Promise<TransactionReceipt> {
-    const blockNumber = await layer.provider.getBlockNumber()
-    const startingBlock = Math.max(blockNumber - this.NUM_BLOCKS_TO_FETCH, 0)
-    const successFilter = {
+    let matches: ethers.providers.Log[] = []
+    const successFilter: ethers.providers.Filter = {
       address: layer.messengerAddress,
       topics: [ethers.utils.id(`RelayedMessage(bytes32)`)],
-      fromBlock: startingBlock,
     }
-    const failureFilter = {
+    const failureFilter: ethers.providers.Filter = {
       address: layer.messengerAddress,
       topics: [ethers.utils.id(`FailedRelayedMessage(bytes32)`)],
-      fromBlock: startingBlock,
     }
-    const successLogs = await layer.provider.getLogs(successFilter)
-    const failureLogs = await layer.provider.getLogs(failureFilter)
-    const logs = successLogs.concat(failureLogs)
-    const matches = logs.filter((log: any) => log.data === msgHash)
+
+    // scan for transaction with specified message
+    while (matches.length === 0) {
+      const blockNumber = await layer.provider.getBlockNumber()
+      const startingBlock = Math.max(blockNumber - this.NUM_BLOCKS_TO_FETCH, 0)
+      successFilter.fromBlock = startingBlock
+      failureFilter.fromBlock = startingBlock
+      const successLogs = await layer.provider.getLogs(successFilter)
+      const failureLogs = await layer.provider.getLogs(failureFilter)
+      const logs = successLogs.concat(failureLogs)
+      matches = logs.filter((log: ethers.providers.Log) => log.data === msgHash)
```

Comment:
Makes me realize that we should emit the msg hash as an indexed event param.

Reply:
Would this be something we should attempt to implement in this PR?

Reply:
For the sake of keeping this PR simple I'd leave it out. But could definitely be worth doing in a separate PR.

Reply:
This change would have to target `regenesis/0.5.0`

Reply:
I think this is an important change as to prevent a massive JSON response that blocks the event loop and causes a memory spike.